### ISSUE TITLE
Unhandled exception. Mono.Linker.MarkException

### ISSUE THREAD

**Issue by mkArtakMSFT** [mono/mono#18651]
_From @americanslon on Thursday, January 30, 2020 10:20:56 PM_

All of the sudden started getting this 

`Unhandled exception. Mono.Linker.MarkException: Error processing method: 'System.IDisposable Microsoft.Extensions.Logging.Logger::BeginScope(TState)' in assembly: 'Microsoft.Extensions.Logging.dll`

I don't even know what else information to provide because it seemingly appeared out of nowhere. Last thing I did before that was connected to azure insights.

I have reinstalled blazor, I have removed the .nuget folder...still the same error. 

_Copied from original issue: dotnet/aspnetcore#18701_

**Comment by mkArtakMSFT** [comment]
_From @mkArtakMSFT on Thursday, January 30, 2020 10:35:00 PM_

Thanks for contacting us.
Please share some more details about what's your scenario, what does your app look like and what did happen for this exception to show up?

**Comment by mkArtakMSFT** [comment]
_From @mkArtakMSFT on Thursday, January 30, 2020 10:35:36 PM_

All that will be helpful for the investigation, which based on this exception seems to be something related to the Mono runtime.

**Comment by mkArtakMSFT** [comment]
_From @americanslon on Thursday, January 30, 2020 10:58:30 PM_

I have a pretty basic client side blazor app served up by and even more basic netcore web app. 
Everything is on the latest preview. 

I was adding serilog to the netcore portion. Added the file sink, then the azure insights sink. Everything is fine.  Connected to  insights. Everything is fine - exception get logged, show up in the insights and files all good. Then literally the next build or so this error. 

As for mono...I don't even know what that is. I am not doing anything with it that blazor isn't already doing.

**Comment by mkArtakMSFT** [comment]
_From @americanslon on Thursday, January 30, 2020 11:19:32 PM_

The only thing that seems to "fix" that error is putting `<BlazorLinkOnBuild >false</BlazorLinkOnBuild>`  in `Blazor.MonoRuntime.Targets` that I fished out from [here](https://github.com/mono/mono/issues/16063)

I put fix in quotes because editing some file that I have never heard of and don't know what it could mean seems like an odd  fix. Also note that the people from that issue don't think it's a mono thing.

**Comment by mkArtakMSFT** [comment]
_From @Madhust on Friday, January 31, 2020 10:46:10 AM_

@mkArtakMSFT I have faced similar exception. 

> C:\Users\MadhuSudhananP\.nuget\packages\microsoft.aspnetcore.blazor.build\3.2.0-preview1.20073.1\targets\Blazor.MonoRuntime.targets(258,5): error : Unhandled exception. Mono.Cecil.ResolutionException: Failed to resolve System.Threading.CancellationToken [D:\Work\temp\TreeShakingTest\TreeShakingTest.csproj]
C:\Users\MadhuSudhananP\.nuget\packages\microsoft.aspnetcore.blazor.build\3.2.0-preview1.20073.1\targets\Blazor.MonoRuntime.targets(258,5): error : ILLink failed with exit code -532462766.

csproj file used in my project is as follows.

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <RazorLangVersion>3.0</RazorLangVersion>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.Blazor" Version="3.2.0-preview1.20073.1" />
    <PackageReference Include="Microsoft.AspNetCore.Blazor.Build" Version="3.2.0-preview1.20073.1" PrivateAssets="all" />
    <PackageReference Include="Microsoft.AspNetCore.Blazor.DevServer" Version="3.2.0-preview1.20073.1" PrivateAssets="all" />
    <PackageReference Include="Microsoft.AspNetCore.Blazor.HttpClient" Version="3.2.0-preview1.20073.1" />
  </ItemGroup>
</Project>
```
Disabling blazor link on build resolves this problem but I don`t want to do that so as I am experimenting on assembly trimming.

Any help would be appreciated. Thanks !!

**Comment by vargaz** [comment]
Looks the same as:
https://github.com/mono/mono/issues/16063

**Comment by kmuthukumar23** [comment]
Hi @mkArtakMSFT  i am also faced the similar exception.

`Severity	Code	Description	Project	File	Line	Suppression State
Error		Unhandled exception. Mono.Linker.MarkException: Error processing method: 'System.Threading.Tasks.Task Microsoft.AspNetCore.Http.Authentication.AuthenticationManager::AuthenticateAsync(Microsoft.AspNetCore.Http.Features.Authentication.AuthenticateContext)' in assembly: 'Microsoft.AspNetCore.Http.Abstractions.dll'	BlazorApp2	C:\Users\MuthukumarKannan\.nuget\packages\microsoft.aspnetcore.blazor.build\3.2.0-preview1.20073.1\targets\Blazor.MonoRuntime.targets	258	
`
It will be resolved when i am disable the blazor linker.

Reproduced Steps:

1. Create the blazor webassembly app
2. Add `Microsoft.AspNetCore.SignalR.Client` latest package.
3. Build the App. 

My Project configuration: 
 `<Project Sdk="Microsoft.NET.Sdk.Web">`

  `<PropertyGroup>`
  `  <TargetFramework>netstandard2.1</TargetFramework>`
    `<RazorLangVersion>3.0</RazorLangVersion>`
  `</PropertyGroup>`

 ` <ItemGroup>`
    `<PackageReference Include="Microsoft.AspNetCore.Blazor" Version="3.2.0-preview1.20073.1" />`
    `<PackageReference Include="Microsoft.AspNetCore.Blazor.Build" Version="3.2.0-preview1.20073.1" PrivateAssets="all" />`
   ` <PackageReference Include="Microsoft.AspNetCore.Blazor.DevServer" Version="3.2.0-preview1.20073.1" PrivateAssets="all" />`
   ` <PackageReference Include="Microsoft.AspNetCore.Blazor.HttpClient" Version="3.2.0-preview1.20073.1" />`
  `  <PackageReference Include="Microsoft.AspNetCore.SignalR.Client" Version="3.1.2" />
  </ItemGroup>`

`</Project> `

SDK Details: 

`C:\Users\MuthukumarKannan>dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.1.200-preview-014883
 Commit:    4e2a0ee959

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17763
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.200-preview-014883\

Host (useful for support):
  Version: 3.1.2
  Commit:  916b5cba26

.NET Core SDKs installed:
  2.1.602 [C:\Program Files\dotnet\sdk]
  2.2.202 [C:\Program Files\dotnet\sdk]
  3.1.100 [C:\Program Files\dotnet\sdk]
  3.1.101 [C:\Program Files\dotnet\sdk]
  3.1.102 [C:\Program Files\dotnet\sdk]
  3.1.200-preview-014883 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download`

Any one help me on this?

**Comment by masoodafar-web** [comment]
The result is not clear?I have the same problem too
`Error processing method: 'System.Threading.Tasks.Task Microsoft.AspNetCore.Http.Authentication.AuthenticationManager::AuthenticateAsync(Microsoft.AspNetCore.Http.Features.Authentication.AuthenticateContext)' in assembly: 'Microsoft.AspNetCore.Http.Abstractions.dll`

**Comment by mkArtakMSFT** [comment]
@lewing I see this is a customer pain point, yet nobody looked at this. Wondering whether there is at least a workaround for this?

**Comment by samfoust** [comment]
Experiencing same issue above when trying to build in release mode with linker. not finding any workarounds at this time. 

`Error processing method: 'System.Threading.Tasks.Task Microsoft.AspNetCore.Http.Authentication.AuthenticationManager::AuthenticateAsync(Microsoft.AspNetCore.Http.Features.Authentication.AuthenticateContext)' in assembly: 'Microsoft.AspNetCore.Http.Abstractions.dll`

`Microsoft.AspNetCore.SignalR.Client`

**Comment by szalapski** [comment]
I am also getting the same issue, in a Blazor WebAssembly Client issue, when building in release configuration.  The only workaround for me is to disable IL Linking with `<BlazorWebAssemblyEnableLinking>false</BlazorWebAssemblyEnableLinking>`

**Comment by diegorod** [comment]
Similar issue myself. Works fine in debug, but get the following with release:

MarkException: Error processing method: 'System.Collections.Generic.IList`1<Sublogix.Model.TableObject> Sublogix.Utility.SalesLogixSchema::GetSchema(System.String,System.Boolean)' in assembly: 'Sublogix.dll'	LogixMobile.Client	C:\Users\drodrigues\.nuget\packages\microsoft.aspnetcore.components.webassembly.build\3.2.0\targets\Blazor.MonoRuntime.targets	326

**Comment by samfoust** [comment]
Yeah I went through and tried a bunch of different framework versions to build on and no luck.
I can build and publish on my local computer in release mode fine, its when I get to Azure Pipelines I run into the issue. Only way around it is to force no linking by dropping
`<BlazorWebAssemblyEnableLinking>false</BlazorWebAssemblyEnableLinking>`
Into the Blazor csproj file. Its looking like its Signalr but I dont know to to get more info around it.

**Comment by szalapski** [comment]
Yes, that matches my situation exactly.  Going to have to disable linking
in production till this is fixed.

**Comment by som-nitjsr** [comment]
I am also facing the same issue

**Comment by redradist** [comment]
I am also have the similar issue:
```console
  Blazor.MonoRuntime.targets(326, 5): Unhandled exception. Mono.Linker.MarkException: Error processing method: 'System.Boolean AutoMapper.ReflectionExtensions::IsDynamic(System.Object)' in assembly: 'AutoMapper.Extensions.ExpressionMapping.dll'
```

**Comment by lewing** [comment]
cc @marek-safar

**Comment by diegofrata** [comment]
I already hit this issue twice with different libraries, the only solution is to disable linking. The feature is essentially broken as sooner or later, as you add features and packages, inevitably one of them will trigger the issue, forcing you to disable the linking.

TL;DR: <BlazorWebAssemblyEnableLinking>false</BlazorWebAssemblyEnableLinking>

**Comment by mezau532** [comment]
had the same Issue when using `Microsoft.AspNetCore.SignalR.Client` version 3.0 and above, the issue ended up being another package I was refrencing `Microsoft.AspNetCore.Http.Abstractions` that only goes up to ASP.NET Core 2.2 which was causing the issue.
